src/queue.c

```diff
@@ -672,23 +677,30 @@ _dispatch_root_queues_init_workq(int *wq_supported)
 		}
 #endif
 	}
-#endif // DISPATCH_USE_LEGACY_WORKQUEUE_FALLBACK || DISPATCH_ENABLE_THREAD_POOL
+#endif // DISPATCH_USE_LEGACY_WORKQUEUE_FALLBACK || DISPATCH_ENABLE_THREAD_POOL || HAVE_INTERNAL_PTHREAD_WORKQUEUE
 #endif // HAVE_PTHREAD_WORKQUEUES
 	return result;
 }
 
 #if DISPATCH_USE_PTHREAD_POOL
 static inline void
 _dispatch_root_queue_init_pthread_pool(dispatch_root_queue_context_t qc,
-		uint8_t pool_size, bool overcommit)
+		int32_t pool_size, bool overcommit)
 {
 	dispatch_pthread_root_queue_context_t pqc = qc->dgq_ctxt;
-	uint32_t thread_pool_size = overcommit ? MAX_PTHREAD_COUNT :
+	int32_t thread_pool_size = overcommit ? DISPATCH_WORKQ_MAX_PTHREAD_COUNT :
 			dispatch_hw_config(active_cpus);
 	if (slowpath(pool_size) && pool_size < thread_pool_size) {
 		thread_pool_size = pool_size;
 	}
 	qc->dgq_thread_pool_size = thread_pool_size;
+	if (qc->dgq_kworkqueue == (void*)(~0ul)) {
+		// when _dispatch_root_queues_init is invoked from
+		// _dispatch_global_queue_poke_slow we need to clear
+		// dgq_pending here because qc->dgq_kworkqueue wasn't
+		// initialized to (~0ul) when we went through _dispatch_global_queue_poke
+		qc->dgq_pending = 0;
```

Comment:
I would do this in `_dispatch_root_queues_init_workq` when we do set it to `~0ull` explaining:

```C
// because the fastpath of _dispatch_global_queue_poke didn't know yet
// that we're using the internal pool implementation we have to undo it setting dgq_pending
```

That would make more sense to me